### PR TITLE
Lighthouse (renamed)

### PR DESCRIPTION
**Pull request by merowin**
This is a prototype for a variant where hidden information is the main quirk. Basically what we discussed in the forum but additionally with "last known information".

The board is hidden, but stones cast rays that illuminate fields for their owner only. Any field that is not illuminated displays its last known information.

I'm trying to build it such that the observer view does not reveal any information to players. I'm not sure how to handle Ko and Self-Capture yet, since players could gain information when they receive an Error and can play another move.

grafik

I've added the possibility to change the background colour of specific fields to MulticolorGridBoard, and a new board component that uses it. I'm not sure if this is good tbh, but I wanted it to test the game class.

Also not sure if fog of war is a good name.

### CODE REVIEW

**Comment by benjaminpjones** [comment]
> I'm trying to build it such that the observer view does not reveal any information to players. I'm not sure how to handle Ko and Self-Capture yet, since players could gain information when they receive an Error and can play another move.

For Ko and self capture, I think it could still be interesting to allow this information leak.  The player must still risk playing a move, so there is no way to purposefully reveal the other player's position.

Of course, there may be other ways as well, such as only allowing moves in the visible area (but this seems much too limiting)

> Also not sure if fog of war is a good name.

I think it's fine.  One concern would be that there seem to be discussions of fog of war go online with various rulesets, but there is no consensus so I think it's fine to have our own.  The whole "board is hidden until stones are played" seems to be the common theme.

**Comment by merowin** [comment]
> For Ko and self capture, I think it could still be interesting to allow this information leak. The player must still risk playing a move, so there is no way to purposefully reveal the other player's position.

I'm thinking to keep the standard behaviour for Ko, but allow self-capture. I want it to be risky to play into the unknown, and that having more visibility is a relatively significant factor.

> I think it's fine. One concern would be that there seem to be discussions of fog of war go online with various rulesets, but there is no consensus so I think it's fine to have our own. The whole "board is hidden until stones are played" seems to be the common theme.

Its a very broad name, also I'm slightly annoyed about causing spaces in the url `govariants.com/variants/fog of war/demo` 😅

**Comment by merowin** [comment]
I'm pretty happy with this now. I decided to rename it to lighthouse, because I feel this name incorporates the ideas well, and it reminds me of the chess variant "bughouse" 😆  
Also updated the visibility rules such that (hopefully) the observer state doesn't reveal any information to players.
Tweaked the colors some:

Blacks perspective:

grafik

Whites perspective:

grafik

The stones in shaded areas were revealed by a "collision".

File: packages/shared/src/lib/grid.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import { Fillable } from "./group_utils";
  * the API should match that of a regular Array.
  */
 export class Grid<T> implements Fillable<CoordinateLike, T> {
-  private arr: Array<T>;
+  protected arr: Array<T>;
```
[the remark this answers is not in the document]
Comment (merowin) [reply packages/shared/src/lib/grid.ts]: Right, thank you for this observation. I've changed it such that Grid.arr is private again.

I was having some problems with making sure that the grid is filled with independent new objects, I tried around a bit and made this unnecessary change.

**Pull request merged**